lib/selinux.c

```diff
@@ -98,6 +98,50 @@ int reset_selinux_file_context (void)
 	return 0;
 }
 
+/*
+ * check_selinux_permit - Checks whether SELinux grants the given operation
+ *
+ *     Returns 0 when permission is granted
+ *                    or something failed but running in permissive mode
+ */
+int check_selinux_permit(const char *perm_name)
+{
+	security_class_t passwd_class_id;
+	access_vector_t permission_id;
+
+	/* caller has checked for (is_selinux_enabled () > 0) already */
+
+	passwd_class_id = string_to_security_class ("passwd");
+	if (0 == passwd_class_id) {
+		fprintf (stderr,
+			    _("%s: can not find SELinux class 'passwd'\n"),
+			    Prog);
+		SYSLOG ((LOG_WARN, "can not find SELinux class 'passwd'"));
+		if (0 != security_getenforce ()) {
+			closelog ();
+			exit (E_SELINUX_ERR);
+		} else {
+			return 0;
+		}
+	}
+
+	permission_id = string_to_av_perm (passwd_class_id, perm_name);
+	if (0 == permission_id) {
+		fprintf (stderr,
+			    _("%s: can not find '%s' SELinux access vector for class 'passwd'\n"),
+			    Prog, perm_name);
+		SYSLOG ((LOG_WARN, "can not find '%s' SELinux access vector for class 'passwd'", perm_name));
+		if (0 != security_getenforce ()) {
+			closelog ();
+			exit (E_SELINUX_ERR);
+		} else {
+			return 0;
+		}
+	}
+
+	return selinux_check_passwd_access (permission_id);
```

Comment:
Wondering if we ought to just convert to direct use of libselinux selinux_check_access().  That would require calling getprevcon() here and passing it as the source and target contexts, but avoids mapping the class twice.  If so, would need/want to also use selinux_set_callback() to set the logging callback and ideally wire that up to audit.